bridge-fungible-token/src/main.sw

```diff
@@ -1,69 +1,186 @@
-        }
-        if (data_length >= 32 + 8) {
-            let num: u64 = input_message_data(msg_idx, 32);
-            storage.data2 = num;
-        }
-        if (data_length >= 32 + 8 + 32) {
-            let big_num: b256 = input_message_data(msg_idx, 32 + 8);
-            storage.data3 = big_num;
-        }
-        if (data_length >= 32 + 8 + 32 + 32) {
-            let address: b256 = input_message_data(msg_idx, 32 + 8 + 32);
-            storage.data4 = ~Address::from(address);
+        require(input_sender.value == LAYER_1_ERC20_GATEWAY, BridgeFungibleTokenError::UnauthorizedUser);
+
+        let message_data = parse_message_data(msg_idx);
+
+        // @review verify `l1_asset` matches hardcoded LAYER_1_TOKEN value
+        require(message_data.l1_asset == ~EvmAddress::from(LAYER_1_TOKEN), BridgeFungibleTokenError::IncorrectAssetDeposited);
+
+        // @review verify `from` matches hardcoded PREDICATE_ROOT value
+        require(message_data.from == ~Address::from(PREDICATE_ROOT), BridgeFungibleTokenError::UnauthorizedUser);
+
+        // @review verify `to` matches hardcoded value TO value
+        require(message_data.to == ~Address::from(TO), BridgeFungibleTokenError::UnauthorizedUser);
+
+        // check that value sent as uint256 can fit inside a u64, else register a refund.
+        let decomposed = decompose(message_data.amount);
+        let amount = ~U256::from(decomposed.0, decomposed.1, decomposed.2, decomposed.3);
+        let l1_amount_opt = amount.as_u64();
+        match l1_amount_opt {
+            Result::Err(e) => {
+                storage.refund_amounts.insert((
+                    message_data.to.value,
+                    message_data.l1_asset,
+                ), amount);
+                // @review emit event (i.e: `DepositFailedEvent`) here to allow the refund process to be initiated?
+            },
+            Result::Ok(v) => {
+                mint_tokens(v, Identity::Address(input_sender));
+                log(555);
+                transfer_tokens(v, contract_id(), input_sender);
+            },
         }
     }
 }
 
-// Implement simple getters for testing purposes
 impl BridgeFungibleToken for Contract {
-    #[storage(read)]
-    fn get_test_counter() -> u64 {
-        storage.counter
+    #[storage(read, write)]
+    // @review can anyone can call this, or only the originator themselves?
+    fn claim_refund(originator: Identity, asset: EvmAddress) {
+        // check storage mapping refund_amounts first
+        // if valid, transfer to originator
+        let inner_value = match originator {
+            Identity::Address(a) => {
+                a.value
+            },
+            Identity::ContractId(c) => {
+                c.value
+            },
+        };
+
+        let amount = storage.refund_amounts.get((
+            inner_value,
+            asset,
+        ));
+        transfer_tokens(amount.as_u64().unwrap(), ~ContractId::from(asset.value), ~Address::from(inner_value));
```

Comment:
A refund shouldn't be transfering any tokens. Refunds are created to refund tokens on Ethereum for when there is an issue processing a message from the gateway (wrong assetId, impossible decimal conversion). This function should be sending an output message back to the gateway similar to a withdrawal.

Reply:
Yeah, I may want to extract the common functionality from both `claim_refund` and `withdraw_to` into a reusable private function.